File: client/java/src/main/java/com/gomelo/CompressionUtil.java
package com.gomelo;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

public class CompressionUtil {

    public enum CompressionType {
        None(0),
        Gzip(1),
        Zlib(2);

        private final int value;
        CompressionType(int value) { this.value = value; }
        public int getValue() { return value; }
    }

    public static class CompressedData {
        public final CompressionType type;
        public final byte[] data;

        public CompressedData(CompressionType type, byte[] data) {
            this.type = type;
            this.data = data;
        }
    }

    public static CompressedData compressGzip(byte[] data) {
        try {
            ByteArrayOutputStream bos = new ByteArrayOutputStream(data.length);
            GZIPOutputStream gzip = new GZIPOutputStream(bos);
            gzip.write(data);
            gzip.close();
            return new CompressedData(CompressionType.Gzip, bos.toByteArray());
        } catch (IOException e) {
            return null;
        }
    }

    public static byte[] decompress(byte[] data) {
        if (data == null || data.length < 2) return data;

        int type = data[0] & 0xFF;
        if (type == CompressionType.None.getValue()) {
            byte[] result = new byte[data.length - 1];
            System.arraycopy(data, 1, result, 0, result.length);
            return result;
        }

        byte[] compressed = new byte[data.length - 1];
        System.arraycopy(data, 1, compressed, 0, compressed.length);

        try {
            if (type == CompressionType.Gzip.getValue()) {
                ByteArrayInputStream bis = new ByteArrayInputStream(compressed);
                GZIPInputStream gzip = new GZIPInputStream(bis);
                ByteArrayOutputStream bos = new ByteArrayOutputStream();
                byte[] buffer = new byte[256];
                int len;
                while ((len = gzip.read(buffer)) != -1) {
                    bos.write(buffer, 0, len);
                }
                return bos.toByteArray();
            }
        } catch (IOException e) {
            return compressed;
        }
        return compressed;
    }

    public static boolean shouldCompress(int dataLength, int threshold) {
        return dataLength >= threshold;
    }
}

File: client/unity/CompressionUtil.cs
using System;
using System.IO;
using System.IO.Compression;

namespace Gomelo.Network
{
    public enum CompressionType : byte
    {
        None = 0,
        Gzip = 1,
        Zlib = 2
    }

    public class CompressedData
    {
        public CompressionType Type { get; private set; }
        public byte[] Data { get; private set; }

        public CompressedData(CompressionType type, byte[] data)
        {
            Type = type;
            Data = data;
        }
    }

    public static class CompressionUtil
    {
        public static CompressedData CompressGzip(byte[] data)
        {
            try
            {
                using var output = new MemoryStream();
                using (var gzip = new GZipStream(output, CompressionLevel.Fastest))
                {
                    gzip.Write(data, 0, data.Length);
                }
                return new CompressedData(CompressionType.Gzip, output.ToArray());
            }
            catch
            {
                return null;
            }
        }

        public static byte[] Decompress(byte[] data)
        {
            if (data == null || data.Length < 2)
                return data;

            CompressionType type = (CompressionType)data[0];
            if (type == CompressionType.None)
                return data.Length > 1 ? data.SubArray(1, data.Length - 1) : Array.Empty<byte>();

            byte[] compressed = new byte[data.Length - 1];
            Array.Copy(data, 1, compressed, 0, compressed.Length);

            try
            {
                if (type == CompressionType.Gzip)
                {
                    using var input = new MemoryStream(compressed);
                    using var gzip = new GZipStream(input, CompressionMode.Decompress);
                    using var output = new MemoryStream();
                    gzip.CopyTo(output);
                    return output.ToArray();
                }
            }
            catch { }

            return compressed;
        }

        public static bool ShouldCompress(int dataLength, int threshold = 512)
        {
            return dataLength >= threshold;
        }
    }
}

public static class ByteArrayExtensions
{
    public static byte[] SubArray(this byte[] data, int offset, int length)
    {
        byte[] result = new byte[length];
        Array.Copy(data, offset, result, 0, length);
        return result;
    }
}